### ISSUE TITLE
Escape ODBC Inserts

### ISSUE THREAD

**Issue by cyberop5** [roadrunnerlenny/etlbox#19]
**Describe the bug**
I'm trying to insert data into an ODBC database (sqlite).  I'm getting an error at DBDestination<>.Wait().
`OdbcException: ERROR [HY000] near "s": syntax error (1)`
The error appears to be the data is being concatenated into a string and executed by the ODBC driver.  The data is not being escaped, nor is the query assembled as a parameterized query.

The problematic line is in OdbcBulkInsertString: `IsAccessDatabase ? $"'{data}' as {destColumnName}" : $"'{data}'";`

**Steps To Reproduce**
1. Create a connection with OdbcConnectionManager.
2. Create a new DBDestination and have it use the OdbcConnection.
3. Create a new source that has a single apostrophe as a value.
4. Link the source to the destination 
5. Create a source file that has an apostrophe as a value (e.g. `"Anne's lunch","You're awesome"`')
6. Call source.Execute() and dbDestination.Wait().

TL;DR - Insert a value into an ODBC database with an apostrophe. It should throw an exception.

**Actual behavior**
`OdbcException: ERROR [HY000] near "s": syntax error (1)`

**Expected behavior**
I would expect the record to be inserted.

**Lazy fix**
Change OdbcBulkInsertString.ValueSql so that it replaces single-quotes `'` with two single-quotes `''`.  Note that this is insecure and may result in a SQL Injection vulnerability.

**Suggested fix**
Change OdbcBulkInsertString so that it creates parameterized queries.  Note that Access databases apparently need special handling: https://docs.microsoft.com/en-us/sql/odbc/microsoft/parameterized-query-limitations?view=sql-server-2017

**Worksarounds**
I could sanitize my input strings by adding a transformer.  This is less than ideal as this behavior seems to be unique to the ODBC connection. 

**Aside** 
I had to create a new CreateTableTask for SQLite support. It doesn't check (yet) if the table already exists.  The current check mechanism was throwing an error. I also changed the primary key from a constraint to `PRIMARY KEY` and re-ordered the output of CreateTableDefinition. I'll have more information soon.

**Comment by roadrunnerlenny** [comment]
Hi,
thank you again for the very detailed bug report. I will have a look into this.

You are suggeesting to use parameterized strings when passing the Sql to the ODBC Connection. I am not quite sure if this will be possible:
One downside of ODBC is that it doesn't support Bulk Insert. The ODBCConnectionManager "fakes" the Bulk Insert by creating a "
```INSERT INTO ... VALUES ( ..) , (..) " if supported. E.g. for access ODBC connections this won't work - there it will be a lot 
I saw that there is also an ADO.NET provider for SqlLite. Perhaps it would help you to have an Sqlite ConnectionMangager? That would probably also help yo

**Comment by roadrunnerlenny** [comment]
Hi,
thank you again for the very detailed bug report. I will have a look into this.

You are sugesting to use parameterized strings when passing the Sql to the ODBC Connection.  I am not quite sure if this will be possible:
One downside of ODBC is that it doesn't support Bulk Insert. (Unless you proof me wrong here) The `ODBCConnectionManager` "fakes" the Bulk Insert by creating a 
```sql
INSERT INTO ... 
VALUES 
( 'a', 'b')
, ( 'd', 'e')
```
As a DBDestination could receive many rows, it would be a bad idea to create one INSERT-statement for every row. 
So I am not quite sure if paramererized queries could really help here? What are your thoughts on that?

Other than that, I saw that there is ADO.NET provider maintained for Sqllite, which also comes with an implementation of Bulk Insert. Would it help you if I could provide you with an SqlLiteConnectionManager? That would mean that you avoid creating an ODBC connector. And it will probably work much faster when bulk inserting data than the workaround for ODBC connections. 

Regarding `CreateTableTask`: Most of the Sql  in the ControlFlow - part of ETLBox was created (and tested) for SqlServer. I'll happy to support more databases. I can have a look at this already.

**Comment by cyberop5** [comment]
I think you're right that ODBC doesn't support bulk insert statements.  

I'm all about parameterized queries.  It'll take a bit more effort to construct the query.  You would want to create the parameterized SQL string 
`INSERT INTO... 
VALUES(?,?), 
(?,?),...`
and build a separate list of parameters (see OdbcParameterCollection.AddWithValue).  I don't think the names used to build the collection matter much.  

Based on my understanding, parameterized queries could be faster than string concatenation because the database may cache the execution plan for the previous batch insert statement and re-use it for subsequent calls.  Non-parameterized queries don't always gain that advantage.  (This may be more pronounced during SELECT statements rather than INSERT statements.)

Nevertheless, inserting raw data via string concatenation is a big no-no.

I'm really enjoying your project. I'm excited for the potential to use it more.  

I predominately use SQL Server, too. I just thought I'd fork the data I'm currently testing it on into a more transient, less resource intensive database until we decide what to do with it.

**Comment by roadrunnerlenny** [comment]
Thanks again for your input. I appreciate your help!

I am going to change the statement in the ODBC connection used for Bulk insert into a parameterized string as you described. Thank you for the good explanation. Probably the execution plan for an insert statement will always be quite simple to be created by the optimizer, but I guess that using parameterized queries is also best-practice and recommended, therefore I agree that's it is a good idea to implement it. I will have a look into it.

In the meanwhile I was able to create a new `SQLiteConnectionManager` based on the ADO.NET provider from SQLite (using System.Data.SQLite). It was easy to implement, so I released version 1.4.0 which does include it. So you can now use the DBDestination with this connection manager, and you will get the BulkInsert provided by SqlLite which should work much faster than using ODBC. Feel free to update via nuget. (Also, it contains the switch to .NET Standard)

I haven't fixed the ODBC issue yet, I will have a look at this in next couple of days. Also I need to fix the `CreateTableTask`. Also I need to check the other tasks as well, as SQLite needs sometimes a different syntax than SqlServer. I will have a look at that hopefully also in the next couple of days. I will keep you updated!

**Comment by roadrunnerlenny** [comment]
An update regarding the Bulk insert using the new SQLiteConnectionManger: I was a little bit to quick - the BulkInsert functionality is not properly implemented yet. So the DBdestination won’t work with this ConnectionManager right now. I will fix that in the next couple of days!

**Comment by roadrunnerlenny** [comment]
Unfortunately, it took me a little longer than expected- I had to do a refactoring of my tests in order to make sure that everything works. 

But now I am pretty confident to say that starting with version 1.5.0 ETLBox can offer full support for SQLite using a native ADO.Net connector. 

I also added the use of parameters when creating the queries in all Connectors. This should fix your issues.
